quint/src/runtime/trace.ts

```diff
@@ -255,28 +256,30 @@ export const newTraceRecorder = (verbosityLevel: number, rng: Rng) => {
       assert(frameStack.length > 0)
       const bottom = frameStack[0]
 
-      let failureOrViolation = true
-      if (outcome.isJust()) {
-        const r = outcome.value.toQuintEx({ nextId: () => 0n })
-        failureOrViolation = r.kind === 'bool' && !r.value
+      const fromResult = (r: Maybe<EvalResult>) => {
+        if (r.isNone()) {
+          return true
+        } else {
+          const rex = r.value.toQuintEx({ nextId: () => 0n })
+          return rex.kind === 'bool' && !rex.value
+        }
       }
 
-      if (failureOrViolation) {
-        if (bestTrace.args.length === 0 || bestTrace.args.length >= bottom.args.length) {
-          // on error, prefer the shorter non-empty trace
-          bestTrace = bottom
-          bestTraceSeed = runSeed
-          bestTrace.result = outcome
-          bestTrace.args = trace
-        }
-      } else {
-        if (bestTrace.args.length <= bottom.args.length) {
-          // on success, prefer the longer trace
-          bestTrace = bottom
-          bestTraceSeed = runSeed
-          bestTrace.result = outcome
-          bestTrace.args = trace
-        }
+      const notOk = fromResult(outcome)
+      const prevNotOk = fromResult(bestTrace.result)
+
+      // override the best trace only if:
+      //  - there is an error, the new trace is shorter, or the old trace is non-error;
+      //  - there is no error, the new trace is longer, and there was no error before.
```

Comment:
I find the old comments about the end-goal of this very important. Can we add that here as well?
```suggestion
      // Prefer short traces for error, and longer traces for non error.
      // Therefore, override the best trace only if:
      //  - there is an error, the new trace is shorter, or the old trace is non-error;
      //  - there is no error, the new trace is longer, and there was no error before.
```